numpy/lib/npyio.py

```diff
@@ -408,31 +408,33 @@ def load(file, mmap_mode=None, allow_pickle=False, fix_imports=True,
 
     pickle_kwargs = dict(encoding=encoding, fix_imports=fix_imports)
 
-    # TODO: Use contextlib.ExitStack once we drop Python 2
     if hasattr(file, 'read'):
         fid = file
         own_fid = False
     else:
         fid = open(os_fspath(file), "rb")
         own_fid = True
 
-    try:
-        # Code to distinguish from NumPy binary files and pickles.
-        _ZIP_PREFIX = b'PK\x03\x04'
-        _ZIP_SUFFIX = b'PK\x05\x06' # empty zip files start with this
-        N = len(format.MAGIC_PREFIX)
-        magic = fid.read(N)
-        # If the file size is less than N, we need to make sure not
-        # to seek past the beginning of the file
-        fid.seek(-min(N, len(magic)), 1)  # back-up
-        if magic.startswith(_ZIP_PREFIX) or magic.startswith(_ZIP_SUFFIX):
-            # zip-file (assume .npz)
-            # Transfer file ownership to NpzFile
-            ret = NpzFile(fid, own_fid=own_fid, allow_pickle=allow_pickle,
-                          pickle_kwargs=pickle_kwargs)
-            own_fid = False
-            return ret
-        elif magic == format.MAGIC_PREFIX:
+    # Code to distinguish from NumPy binary files and pickles.
+    _ZIP_PREFIX = b'PK\x03\x04'
+    _ZIP_SUFFIX = b'PK\x05\x06' # empty zip files start with this
+    N = len(format.MAGIC_PREFIX)
+    magic = fid.read(N)
+    # If the file size is less than N, we need to make sure not
+    # to seek past the beginning of the file
+    fid.seek(-min(N, len(magic)), 1)  # back-up
+    if magic.startswith(_ZIP_PREFIX) or magic.startswith(_ZIP_SUFFIX):
+        # zip-file (assume .npz)
+        # Potentially transfer file ownership to NpzFile
+        ret = NpzFile(fid, own_fid=own_fid, allow_pickle=allow_pickle,
+                      pickle_kwargs=pickle_kwargs)
+        return ret
+
+    with contextlib.ExitStack() as stack:
+        if own_fid:
```

Comment:
I don't think this came out nicer than the original code but I'm leaving it here to show what I attempted.

The main difficulty is having to transfer file ownership to NpzFile.
I tried a number of variations on stack.pop_all() there but none of them came out nicer.